### DISCUSSION TITLE
How to use Conv1d in Julia

### DISCUSSION THREAD

**Discussion by bjoern234** [apache/incubator-mxnet#17383] [Q&A]
Hi!
I've tried to translate the Conv1d Python Keras/TF example from [https://machinelearningmastery.com/cnn-models-for-human-activity-recognition-time-series-classification/](https://machinelearningmastery.com/cnn-models-for-human-activity-recognition-time-series-classification/) into MXNet.jl. 

I was able to train the model but with poor results compared to the Keras version. I'm not sure what I've done wrong. I assume something is wrong with the input shape or the model. I'm a little bit confused how to use the mx.ArrayDataProvider. As can you see the accuracy after 20 epochs is just ~20% instead of ~90% using Keras. I've tried different input shapes but without success :(

The dataset can be downloaded from:
[https://archive.ics.uci.edu/ml/machine-learning-databases/00240/UCI%20HAR%20Dataset.zip](https://archive.ics.uci.edu/ml/machine-learning-databases/00240/UCI%20HAR%20Dataset.zip)

Thanks in advance!
```
using MXNet
using DelimitedFiles


path_dataset = "/Users/bjoern/code/tests/HARDataset"

# load a single file as a numpy array
function load_file(file_path)
    dat = readdlm(file_path)
    return dat
end


# load a list of files and return as a 3d numpy array
function load_group(filenames, path2csv)
    loaded = []
    for name in filenames
        data = load_file(path2csv * name)
        push!(loaded, data)
    end
    # stack group so that features are the 3rd dimension
    loaded = cat(loaded...; dims=3)
    return loaded
end



# load a dataset group, such as train or test
function load_dataset_group(group, path_dataset)
    filepath = joinpath(path_dataset, group, "Inertial_Signals/")
    # load all 9 files as a single array
    filenames = []
    # total acceleration
    push!(filenames, "total_acc_x_" * group * ".txt", "total_acc_y_" * group * ".txt", "total_acc_z_" * group * ".txt")
    # body acceleration
    push!(filenames, "body_acc_x_" * group * ".txt", "body_acc_y_" * group * ".txt", "body_acc_z_" * group * ".txt")
    # body gyroscope
    push!(filenames, "body_gyro_x_" * group * ".txt", "body_gyro_y_" * group * ".txt", "body_gyro_z_" * group * ".txt")
    # load input data
    X = load_group(filenames, filepath)
    # load class output
    y = load_file(path_dataset * "/" * group * "/y_" * group * ".txt")
    return X, y
end



# load the dataset, returns train and test X and y elements
function load_dataset(path_dataset)
    # load all train
    trainX, trainy = load_dataset_group("train", path_dataset)
    println(size(trainX), size(trainy))
    # load all test
    testX, testy = load_dataset_group("test", path_dataset)
    println(size(testX), size(testy))
    println(size(trainX), size(trainy), size(testX), size(testy))
    return trainX, trainy, testX, testy
end


# load data
trainX, trainy, testX, testy = load_dataset(path_dataset)

trainX_ = permutedims(trainX, (2,3,1))
testX_ = permutedims(testX, (2,3,1))

trainy_ = reshape(trainy,(7352,))
testy_ = reshape(testy, (2947,))

train_data_provider = mx.ArrayDataProvider(:data => trainX_, :label => trainy_, batch_size=32, shuffle=true)
validation_data_provider = mx.ArrayDataProvider(:data => testX_, :label => testy_,batch_size=32)


arch = @mx.chain mx.Variable(:data) =>
        #mx.Embedding(2,1) =>
        mx.Convolution(kernel=(3,), num_filter=64) =>
        mx.Activation(act_type=:relu) =>
        mx.Convolution(kernel=(3,), num_filter=64) =>
        mx.Activation(act_type=:relu) =>
        mx.Dropout(p=0.5) =>
        mx.Pooling(pool_type=:max, kernel=(2,)) =>
        mx.Flatten() =>
        mx.FullyConnected(num_hidden=100) =>
        mx.Activation(act_type=:relu) =>
        mx.FullyConnected(num_hidden=6) =>
        mx.SoftmaxOutput(mx.Variable(:label))

nnet = mx.FeedForward(arch, context = mx.cpu())
mx.fit(nnet, mx.ADAM(), train_data_provider, eval_data = validation_data_provider, n_epoch = 20, callbacks = [mx.speedometer()]);


 Info: Start training on Context[CPU0]
[ Info: Initializing parameters...
[ Info: Creating KVStore...
[ Info: TempSpace: Total 0 MB allocated on CPU0
[ Info: Start training...
[ Info: Speed: 1303.57 samples/sec
[ Info: Speed: 1364.40 samples/sec
[ Info: Speed: 1282.10 samples/sec
[ Info: Speed: 1306.47 samples/sec
[ Info: == Epoch 001/050 ==========
[ Info: ## Training summary
[ Info:           accuracy = 0.2117
[ Info:               time = 5.6121 seconds
[ Info: ## Validation summary
[ Info:           accuracy = 0.0413
[ Info: Speed: 1378.38 samples/sec
[ Info: Speed: 1350.01 samples/sec
[ Info: Speed: 1353.40 samples/sec
[ Info: Speed: 1368.58 samples/sec
[ Info: == Epoch 002/050 ==========
[ Info: ## Training summary
[ Info:           accuracy = 0.1776
[ Info:               time = 5.3959 seconds
[ Info: ## Validation summary
[ Info:           accuracy = 0.2530
[ Info: Speed: 1341.65 samples/sec
[ Info: Speed: 1366.51 samples/sec
[ Info: Speed: 1254.77 samples/sec
[ Info: Speed: 1205.79 samples/sec
.
.
.
[ Info: == Epoch 019/050 ==========
[ Info: ## Training summary
[ Info:           accuracy = 0.1822
[ Info:               time = 6.5816 seconds
[ Info: ## Validation summary
[ Info:           accuracy = 0.1411
[ Info: Speed: 1195.42 samples/sec
[ Info: Speed: 1235.22 samples/sec
[ Info: Speed: 1226.30 samples/sec
[ Info: Speed: 1240.68 samples/sec
[ Info: == Epoch 020/050 ==========
[ Info: ## Training summary
[ Info:           accuracy = 0.1792
[ Info:               time = 6.0004 seconds
[ Info: ## Validation summary
[ Info:           accuracy = 0.1788
[ Info: Speed: 1338.49 samples/sec
```
